### DISCUSSION TITLE
What could cause problems with opening and saving a copy of an XFA PDF file?

### DISCUSSION THREAD

**Discussion by raemery** [pikepdf/pikepdf#541] [Q&A]
Hi,

I'm attempting to write a script that will populate a government form that uses XFA. I have been able to use pikepdf successfully to extract and view information in the form, but each time I try to populate and save date, the resulting pdf doesn't work any more. In the process of trying to isolate the issue, I eventually got to the point where simply opening the file, reading it, and saving a copy to a different file name would reproduce the error. Any thoughts on what might be going on to cause this, or ways around it?

Here's the code I'm using to open and save a copy:
```
import pikepdf

# Open the PDF file
with pikepdf.Pdf.open('forms/NFFBAR.pdf') as pdfData:
    # Save the PDF file
    pdfData.save('forms/NFFBAR_copy.pdf')
```

Here's a link to the original PDF (click on the "Prepare" button to download):
[https://bsaefiling.fincen.treas.gov/NoRegFBARFiler.html](https://bsaefiling.fincen.treas.gov/NoRegFBARFiler.html)

Comparing the PDF copy to the original PDF, the buttons no longer function, and there is a  "#USER_SITE_ERROR#' message displayed on the first page under the "Sign the Form" button.

Inside the PDF file, template.xml holds the javascript for the buttons and also the error message.  All the data entered by the user is stored in datasets.xml.

**Comment by jbarlow83** [comment]
XFA is nonstandard, deprecated, and the tools and documentation were only available to Adobe's government customers or perhaps some large businesses. I believe there is some sort of cryptographic check to ensure only Acrobat can edit it.

If you email me I can put you in touch with someone I know who has worked on XFA and may be able to help, although I'm not sure how successful his venture was. Your best bet may be to do something awful like use AutoHotkey to automate Acrobat.

**Reply by raemery** [reply]
First, thank you very much for the quick reply. I'm a little embarrassed to say this, but I'm not sure where I can email you on GitHub or find your email, but I would love to follow up.

For what it's worth, I did find a commercial product written in PHP that can edit some but not all of the XML data in the same form and successfully save a working copy: https://www.setasign.com/products/setapdf-formfiller/demos/php-xfa-demo/#p-358

**Reply by jbarlow83** [reply]
It's in pyproject.toml. Sorry, just trying to avoid the spam....
